### PR TITLE
Discontinue relationship

### LINKED ISSUE #8: Create a mutation to delete relationship record when rock declines
Takes two arguments:
- rock-and-pebble record ID
- reason (string)

Deletes rock-and-pebble relationship with that ID, return all that rock's updated rocks and pebbles data.

Side effect:
- Sends polite, regretful email to pebble that their request was declined, with the reason that was given.

### PR DESCRIPTION
**Pull request by rer7891**
### Sidebar Checklist

- [x] Request reviewers
- [x] Assign yourself and other contributors
- [x] Link to project

### Issues Resolved
Resolves #8 

### Problem Addressed
Needed a mutation to discontinue a rock and pebble relationship
### Solution Implemented
added a new mutation and tested mutation
### Other Notes
added helper methods in user model to return only active pebbles and rocks

### CODE REVIEW

**Review by msimon42 — APPROVED** [review]
Looks good to me.

File: app/graphql/mutations/rock_and_pebbles/discontinue_rock_pebble_relationship.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+          rock_and_pebble.update(active: false)
+          rock_and_pebble
+        end
+
+        private
+
+        def send_pebble_email(rock_and_pebble, attributes)
+          user = User.find(attributes[:user_id])
+          if user.id == rock_and_pebble.rock_id
```
Comment (DanielEFrampton) [inline app/graphql/mutations/rock_and_pebbles/discontinue_rock_pebble_relationship.rb]: Small optimization suggestion: instead of the User lookup on line 20, could we change line 21 to use the `attributes[:user_id]` value?
```suggestion
          if attributes[:user_id] == rock_and_pebble.rock_id
```

File: app/graphql/mutations/rock_and_pebbles/discontinue_rock_pebble_relationship.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+            reason = attributes[:reason]
+            info = rock_and_pebble.rock_pebble_info
```
Comment (DanielEFrampton) [inline app/graphql/mutations/rock_and_pebbles/discontinue_rock_pebble_relationship.rb]: Another small optimization: because this is the same in both the `if` and `else`, this could be moved up to the top of the method and lines 26-27 could be deleted

File: spec/graphql/mutations/rock_and_pebble/decline_rock_pebble_spec.rb [hunk 3]
```diff
@@ -7,6 +7,7 @@
       @user_2 = create :user
       @user_3 = create :user
       @user_1.pebbles << [@user_2, @user_3]
+      RockAndPebble.last.update(active: true)
```
Comment (DanielEFrampton) [inline spec/graphql/mutations/rock_and_pebble/decline_rock_pebble_spec.rb:10]: Question about this test, not this specific line; would it possible to add an assertion confirming that the email has been sent (or at least added to the pipeline/queue or whatever it's called)? I vaguely recall being able to do that on one of our mod 3 projects but can't remember which.

File: spec/graphql/mutations/rock_and_pebble/discontinue_rock_pebble_spec.rb [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+                              }
+                  }
+                }
+              GQL
+
+      @rock = @rock_and_pebble.rock
+    end
+
+    it 'deactivates the rock and pebble relationship' do
+      expect(@rock_and_pebble.reload.active).to eq(true)
```
Comment (DanielEFrampton) [inline spec/graphql/mutations/rock_and_pebble/discontinue_rock_pebble_spec.rb:43]: I don't think I've ever seen `reload` before, that's cool!

File: spec/graphql/mutations/rock_and_pebble/discontinue_rock_pebble_spec.rb [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+      post '/graphql', params: {query: @query1}
+      result = JSON.parse(response.body)["data"]["rock_and_pebble"]
+
```
Comment (DanielEFrampton) [inline spec/graphql/mutations/rock_and_pebble/discontinue_rock_pebble_spec.rb:47]: As I mentioned above, I wonder if it'd be possible/helpful to assert that the email got added to the outbound mail queue somewhere in this test

File: app/models/user.rb [hunk 6]
```diff
@@ -38,4 +38,16 @@ def user_total_hours_mentored
...
+  where('users.id IN (SELECT rock_and_pebbles.pebble_id FROM users join rock_and_pebbles
+         on rock_and_pebbles.rock_id = users.id WHERE rock_and_pebbles.active = true AND 
+         rock_and_pebbles.rock_id = ?)', id)
+  end 
+
+  def self.active_rocks(id)
+    where('users.id IN (SELECT rock_and_pebbles.rock_id FROM users join rock_and_pebbles
+    on rock_and_pebbles.rock_id = users.id WHERE rock_and_pebbles.active = true AND 
+    rock_and_pebbles.pebble_id = ?)', id)
+  end
```
Comment (DanielEFrampton) [inline app/models/user.rb]: Nice sub-querying!

I wonder if, because you have to pass in the ID for a specific user anyway, these would look slightly cleaner as instance methods? (I think `id` would be available by virtue of being another instance method at the same scope, but if not `self.id` would grab it)

```suggestion
  def active_pebbles
  where('users.id IN (SELECT rock_and_pebbles.pebble_id FROM users join rock_and_pebbles
         on rock_and_pebbles.rock_id = users.id WHERE rock_and_pebbles.active = true AND 
         rock_and_pebbles.rock_id = ?)', id)
  end 

  def active_rocks
    where('users.id IN (SELECT rock_and_pebbles.rock_id FROM users join rock_and_pebbles
    on rock_and_pebbles.rock_id = users.id WHERE rock_and_pebbles.active = true AND 
    rock_and_pebbles.pebble_id = ?)', id)
  end 
```

File: app/graphql/types/rock_and_pebble_type.rb [hunk 7]
```diff
@@ -9,12 +9,20 @@ class RockAndPebbleType < Types::BaseObject
...
     def my_rocks
       user = User.find(object.pebble_id)
-      user.rocks
+      User.active_rocks(user.id)
     end
 
     def my_pebbles
       user = User.find(object.rock_id)
-      user.pebbles
+      User.active_pebbles(user.id)
```
Comment (DanielEFrampton) [inline app/graphql/types/rock_and_pebble_type.rb]: (See comment below on `user.rb`)

Comment (DanielEFrampton) [reply app/graphql/types/rock_and_pebble_type.rb]: If you decide to keep `active_pebbles` as a User class method, I think you could shorten this to skip the User lookup on line 16 for a small optimization:

```
def my_rocks
  User.active_rocks(object.pebble_id)
end

def my_pebbles
  User.active_pebbles(object.rock_id)
end
```

**Review by DanielEFrampton — APPROVED** [review]
This looks great, Becky! I left some optimization suggestions but nothing critical. Thanks for finishing this up!

File: spec/graphql/mutations/rock_and_pebble/decline_rock_pebble_spec.rb (hunk 3 above)
Comment (DanielEFrampton) [reply spec/graphql/mutations/rock_and_pebble/decline_rock_pebble_spec.rb:10]: (We could also use stubs to verify that the message gets where it's supposed to if we wanted to isolate this test from that whole process)

**Comment by DanielEFrampton** [comment]
Also resolves #9 I believe

File: app/graphql/mutations/rock_and_pebbles/discontinue_rock_pebble_relationship.rb (hunk 1 above)
Comment (rer7891) [reply app/graphql/mutations/rock_and_pebbles/discontinue_rock_pebble_relationship.rb]: Done!

File: app/graphql/mutations/rock_and_pebbles/discontinue_rock_pebble_relationship.rb (hunk 2 above)
Comment (rer7891) [reply app/graphql/mutations/rock_and_pebbles/discontinue_rock_pebble_relationship.rb]: Haha I thought of this after pushing and was being lazy! Changes pushed up

File: spec/graphql/mutations/rock_and_pebble/decline_rock_pebble_spec.rb (hunk 3 above)
Comment (rer7891) [reply spec/graphql/mutations/rock_and_pebble/decline_rock_pebble_spec.rb:10]: While I think this is a good idea I have not implemented it yet. We don't do this for any other mutations that send out an email so perhaps this should be it's own ticket? To update all necessary tests?

File: app/models/user.rb (hunk 6 above)
Comment (rer7891) [reply app/models/user.rb]: Made an instance method.

File: spec/graphql/mutations/rock_and_pebble/decline_rock_pebble_spec.rb (hunk 3 above)
Comment (DanielEFrampton) [reply spec/graphql/mutations/rock_and_pebble/decline_rock_pebble_spec.rb:10]: Yeah, that makes sense to me too. I'll make an issue for it

**Comment by DanielEFrampton** [comment]
@rer7891 I can't remember what our previous procedure was, but feel free to merge this whenever you feel it's ready
